### ISSUE TITLE
`TypeVar` related errors `invalid-parameter-default` and `invalid-method-override`

### ISSUE THREAD

**Issue by fbuccioni** [astral-sh/ty#3052]
### Summary

Hi, first of all thanks for the awesome project.

I have the following issues related to `TypeVar` I think maybe is because it needs to read the bound parameter.

```py
from typing import Literal, Any, Generic
from typing_extensions import TypeVar


OptionalStructBase = dict[str, Any] | None
OptionalStruct = TypeVar('OptionalStruct', default=None, bound=OptionalStructBase)
Type = TypeVar('Type', bound=Any, default=None)


class DataHandlerBase():
    # The first one, the default value for a variable I found a related issue
    # but I don't know if in the internals is the same or not.
    def apply(struct: OptionalStruct = None) -> Any:
        ...
    
    def struct(**struct: OptionalStructBase) -> Any:
        ...

class DataConcrete(DataHandlerBase, Generic[OptionalStruct, Type]):
    # Inheritance problems with generic bounded types
    def struct(**struct: OptionalStruct) -> None:
        ...

```

https://play.ty.dev/95a472fc-3d38-4ce1-93b1-f6bf6511198b

Thanks in advance.

### Version

`0.0.23`

**Comment by sharkdp** [comment]
Thank you for reporting this.

Here's a smaller example that demonstrates the same problem:
```py
# ty: Default value of type `None` is not assignable to annotated parameter type `T@f`
def f[T = int | None](x: T = None) -> T:
    return x
```

It's interesting to me that mypy and pyrefly also emit an error here (pyright does not). It seems to me that this should clearly be allowed? And the call expression `f()` should be of type `None`.

**Comment by AlexWaygood** [comment]
The `invalid-parameter-default` part of this is https://github.com/astral-sh/ty/issues/592, though I agree that it's even clearer that this should be allowed when the type parameter has a default.

That leaves the second issue mentioned in the OP, which was an `invalid-method-override` diagnostic where none was expected. The OP's example of this was:

```py
from typing import Any, Generic, TypeVar

OptionalStructBase = dict[str, Any] | None
OptionalStruct = TypeVar('OptionalStruct', default=None, bound=OptionalStructBase)
Type = TypeVar('Type', bound=Any, default=None)

class DataHandlerBase:
    def struct(**struct: OptionalStructBase) -> Any: ...

class DataConcrete(DataHandlerBase, Generic[OptionalStruct, Type]):
    def struct(**struct: OptionalStruct) -> None: ...
```

and a more minimal reproducer of the issue is

```py
class Foo: ...

class Parent:
    def method(**parameter: Foo): ...

class Child[T: Foo](Parent):
    def method(**parameter: T): ...
```

But I think this is a true positive. In my minimal example here, `Parent.method` accepts any keyword argument as long as the value passed to that argument is an instance of `Foo`. But concrete specializations of `Child` (which are all subtypes of `Parent`) will not necessarily accept instances of `Foo` passed by keyword argument to the `method` method:

```py
class FooSub(Foo): ...

# This is a valid specialization of `Child`,
# and `ConcreteChild` must be considered a subtype of `Parent`,
# but the `method` method on instances of `ConcreteChild`
# accepts a narrower type than the `method` method on `Parent`,
# violating the Liskov Substitution Principle
ConcreteChild = Child[FooSub]
```
